src/lib/jobrunner.rs

```diff
@@ -0,0 +1,409 @@
+                };
+                let stdout_fd = if job.stdout_hup {
+                    -1
+                } else {
+                    job.child.stdout.as_ref().unwrap().as_raw_fd()
+                };
+                (stderr_fd, stdout_fd)
+            } else {
+                (-1, -1)
+            };
+            self.pollfds[i * 2] = PollFd::new(stderr_fd, PollFlags::POLLIN);
+            self.pollfds[i * 2 + 1] = PollFd::new(stdout_fd, PollFlags::POLLIN);
+        }
+        self.pollfds[self.snare.config.maxjobs * 2] =
+            PollFd::new(self.snare.event_read_fd, PollFlags::POLLIN);
+    }
+
+    /// If the user has specified an email address, send the contents of
+    fn sendemail(&self, mut file: &File) {
+        if let Some(ref toaddr) = self.snare.config.email {
+            let mut buf = Vec::new();
+            buf.extend("Subject: snare error\n\n".as_bytes());
+            file.seek(SeekFrom::Start(0)).ok();
+            file.read_to_end(&mut buf).ok();
+
+            let fromea = match EmailAddress::new(format!("{}@{}", username(), hostname())) {
+                Ok(ea) => Some(ea),
+                Err(_) => None,
+            };
+
+            match EmailAddress::new(toaddr.to_string()) {
+                Ok(toea) => {
+                    let email = SendableEmail::new(
+                        Envelope::new(fromea, vec![toea]).unwrap(),
+                        "na".to_string(),
+                        buf,
+                    );
+
+                    let mut sender = SendmailTransport::new();
+                    if let Err(e) = sender.send(email) {
+                        self.warning(&format!("Couldn't send email: {:?}", e));
+                    }
+                }
+                Err(_) => self.warning(&format!("Invalid To: email address {}", toaddr)),
+            }
+        }
+    }
+
+    /// Log a warning to the user. This might be to stderr, a log file, or email.
+    fn warning(&self, msg: &str) {
+        eprintln!("Warning: {}\n", msg);
+    }
+}
+
+struct Job {
+    child: Child,
+    /// This TempDir will be dropped, and its file system contents removed, when this Job is dropped.
+    _tempdir: TempDir,
+    /// We are responsible for manually cleaning up the JSON file stored in `json_path`.
+    json_path: PathBuf,
+    /// The file to which we write combined stderr/stdout.
+    stderrout_file: File,
+    /// Has the child process's stderr been closed?
+    stderr_hup: bool,
+    /// Has the child process's stdout been closed?
+    stdout_hup: bool,
+}
+
+fn set_nonblock(fd: RawFd) -> Result<(), Box<dyn Error>> {
+    let mut flags = fcntl(fd, FcntlArg::F_GETFL)?;
+    flags |= OFlag::O_NONBLOCK.bits();
+    fcntl(
+        fd,
+        FcntlArg::F_SETFL(unsafe { OFlag::from_bits_unchecked(flags) }),
+    )?;
+    Ok(())
+}
+
+pub(crate) fn attend(snare: Arc<Snare>) -> Result<(), Box<dyn Error>> {
+    let mut rn = JobRunner::new(snare)?;
+    thread::spawn(move || rn.attend());
```

Comment:
Do we get an appropriate error message if this spawn fails? I guess you have to check something in the join handle?

Reply:
Don't you check for error when `join()`ing via the handle?

Reply:
There is no `join` here, as that would block the HTTP thread!

Reply:
Oh. Hrm.

So supposing that thread panics *or* fails to spawn (i think the cases are indistinguishable), would another part of the program notice and print an error?

If not, that's bad. Perhaps we need to store the handle somewhere and check for error at such a time when it is safe (in that it won't block anything).

Reply:
As i said, I don't know how to do this in the current API...

Reply:
Indeed. This is annoying.

I tried making `attend()` panic to see what happens:
```
     Running `target/debug/snare -p 3333 -r /tmp -s /tmp/s`
thread '<unnamed>' panicked at 'fish', src/lib/jobrunner.rs:65:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

And then the program hangs.

Oh well.

Reply:
I think we have to accept this as a limitation of the API, at least for now. Maybe we'll work out how to do it in the future.